app/auth/gitlab_auth.py

```diff
@@ -40,198 +36,80 @@
     url_for,
 )
 
-from .web import JWT_ALGORITHM, get_key_for_user
+from .oauth_provider_app import GitLabProviderApp
+from .utils import (
+    get_redis_key_from_session,
+    get_redis_key_from_token,
+    handle_login_request,
+    handle_token_request,
+)
+from .oauth_client import RenkuWebApplicationClient
+
+GL_SUFFIX = "gl_oauth_client"
 
 blueprint = Blueprint("gitlab_auth", __name__, url_prefix="/auth/gitlab")
 
 
+# Note: GitLab oauth tokens do NOT expire per default
+# See https://gitlab.com/gitlab-org/gitlab/-/issues/21745
+# The documentation about this is wrong
+# (https://docs.gitlab.com/ce/api/oauth2.html#web-application-flow)
+
+
 class GitlabUserToken:
     def __init__(self, header_field="Authorization", header_prefix="Bearer "):
         self.header_field = header_field
         self.header_prefix = header_prefix
 
     def process(self, request, headers):
-        from .. import store
 
         m = re.search(
             r"bearer (?P<token>.+)", headers.get("Authorization", ""), re.IGNORECASE
         )
         if m:
             # current_app.logger.debug('outgoing headers: {}'.format(json.dumps(headers))
             access_token = m.group("token")
-            decodentoken = jwt.decode(
-                access_token,
-                current_app.config["OIDC_PUBLIC_KEY"],
-                algorithms=JWT_ALGORITHM,
-                audience=current_app.config["OIDC_CLIENT_ID"],
-            )
-
-            gl_token = store.get(get_key_for_user(decodentoken, "gl_access_token"))
-            headers[self.header_field] = "{}{}".format(
-                self.header_prefix, gl_token.decode()
+            gitlab_oauth_client = current_app.store.get_oauth_client(
+                get_redis_key_from_token(access_token, key_suffix=GL_SUFFIX)
             )
-            headers[
-                "Renku-Token"
-            ] = access_token  # can be needed later in the request processing
+            if gitlab_oauth_client:
+                headers[self.header_field] = "{}{}".format(
```

Comment:
Not sure which is the repository standard for string formatting, but since it specifies python 3.7+ as minimal version, you could checkout f-strings, it would make things a bit more readable.

Reply:
Agreed: https://github.com/SwissDataScienceCenter/renku-gateway/issues/234